components/encryption/export/Cargo.toml

```diff
@@ -0,0 +1,39 @@
+[package]
+name = "encryption_export"
```

Comment:
why we need this extra export crate? We can have `encryption` crate depend directly on `cloud` crate.

Reply:
We could. The normal approach is that a trait is defined in a crate and it is implemented for the standard library and some other light-weight crates. Other crates implementing the interface import the interface. Reversing this means that crates that rely on encryption are now being imported by encryption instead. 

It would be odd to see the encryption lib depend on cloud-aws and thus the AWS SDK. The AWS SDK libraries here will soon include S3. We could make a separate AWS cloud crate for encryption and storage, but still we have our encryption library depending on AWS KMS even though it shouldn't know anything about this. Normally dependencies flow in the opposite direction.

It would make sense for cloud-aws to import encryption for KMS encryption. We are choosing to do extra work to de-couple these through a cloud interface. Would we follow this pattern for other crates? We are only allowed to by the compiler if we go to the extra trouble of creating additional traits that allow for de-coupling.

I am open to being convinced that this is a pragmatic choice. However, just because we can now reverse the dependency flow doesn't necessarily mean we should.

Reply:
I'm seeing your point that encryption crate need not depend on cloud-aws. But I still want to get rid of encryption-export which is an extra indirection. How about we do this?

* split `cloud` crate into `cloud-trait` and `cloud`. `cloud-trait` contain the traits for KMS and blob storage. `cloud` contain instantiation of cloud providers that you are putting in `xxx-export` crates.
* `encryption` crate depend on `cloud-trait`.
* Move `encryption::KMSConfig` config struct into the `cloud-trait` crate (becomes `cloud-trait::KMSConfig` probably). `encryption::KMSConfig` become an alias of `cloud-trait::KMSConfig`.
* in the server layer where `encryption::DataKeyManager` is created, the KMS provider is constructed and passed into the `DataKeyManager`'s ctor. Something like:

```
// in cmd/src/server.rs
let encryption_config = tikv_config.security_encryption;
let kms_provider = if let MasterKeyConfig::KMS { kms_config } = encryption_config.master_key {
  Some(cloud::kms_provider_from_config(kms_config))
} else {
  None
};
let encryption_key_manager = encryption::DataKeyManager::new(..., kms_provider);
```

What do you think?

Reply:
This would be fine if we were sure the server will always be the only code using this. But we do already have tikv-ctl using this so it already gets repetitive. It might work okay for encryption which currently has few end users, but this is not a broad pattern that can be followed. external storage already has backup and sst_importer as end users in the code base and could end up with more end users.

I think that the "extra indirection" has its purpose in this general pattern to differentiate between libraries that implement the interface and end users that constructor it and call its methods. It simplifies usage to only have a restricted subset of encryption available for import (as encryption_export).

Reply:
I don't see encryption crate or the external_storage crate being use in say 10 places in the foreseeable future. The added code to pass the param is almost just a one liner. Change both of these crate to accept KMSProvider and BlobStorageProvider as param also make them more testable - for dependency injection.

Reply:
I changed things so that neither crate depends on each other and instead they are bound together in the export crate via newtype delegation from cloud to encryption. I think this is a nice organization.

Reply:
Again I think your suggestions are pragmatic enough for the situation we have right now. However, I am much more interested in establishing patterns that we can use everywhere, are easy to follow, and don't have too much overhead.